### PR TITLE
fix #907: add support for std::string_view for C++17

### LINKED ISSUE #907: better std::experimental::string_view support
std::experimental::string_view has operator<< overload for std::ostream but still printed as collection of chars. The only way to make gtest print string_view value as string is add PrintTo into std::experimental::fundamentals_v1 namespace which is not really great practice.

### PR DESCRIPTION
**Pull request by dermojo**
Implementation is identical to absl::string_view.

### CODE REVIEW

**Comment by gennadiycivil** [comment]
Apologies for the long wait. googletest can conditionally depend on absl:: so absl:string_view would be available

**Comment by KindDragon** [comment]
@gennadiycivil This PR for **stl**:string_view. Please check PR again

**Comment by gennadiycivil** [comment]
Aplogies for not explaining this clear-er. absl::string_view IS std:string_view under C++ 17
See https://abseil.io/blog/20170926-welcome-to-abseil. 
Thanks again

**Comment by gennadiycivil** [comment]
... and https://abseil.io/about/design/dropin-types

**Comment by dermojo** [comment]
Thanks for taking the time to explain. Maybe I should have done the same.
The fact that both types are identical doesn't seem to help - when using std::string_view with GTest (e.g. in an ASSERT_EQ), it will be printed as a plain array of int's. Maybe this works when absl is actually used and included, but it doesn't if not.

Sample code:
```
#include <gtest/gtest.h>
#include <string_view>

TEST(Foo, StringView)
{
   std::string_view s1 = "Hello";
   std::string_view s2 = "World";
   ASSERT_EQ(s1, s2);
}
```
This results in the following test error:
```
ex.cpp:8: Failure
      Expected: s1
      Which is: { 'H' (72, 0x48), 'e' (101, 0x65), 'l' (108, 0x6C), 'l' (108, 0x6C), 'o' (111, 0x6F) }
To be equal to: s2
      Which is: { 'W' (87, 0x57), 'o' (111, 0x6F), 'r' (114, 0x72), 'l' (108, 0x6C), 'd' (100, 0x64) }
```

But it should look like this (same output as using std::string):
```
ex.cpp:8: Failure
      Expected: s1
      Which is: "Hello"
To be equal to: s2
      Which is: "World"
```

**Comment by dermojo** [comment]
@gennadiycivil Bump (not sure if my latest comment triggered a notification)

**Comment by gennadiycivil** [comment]
@dermojo You are correct, I have not seen your previous notification. I ll re-open this PR , I have to apologize in advance because it will be some time before I can get back to it, we have higher priorities currently with 1.8.x release and cleanup work.

**Comment by dermojo** [comment]
Thanks! Don't worry, take your time.

**Comment by Jonny007-MKD** [comment]
@dermojo: You could instead implement this PR for absl::string_view, so that anybody who is using absl::string_view with pre-C++17 compilers would also have nice output in gtest, whereas everybody with C++17 would in fact be having code for std::string_view (see https://abseil.io/about/design/dropin-types)

**Comment by dermojo** [comment]
@Jonny007-MKD: Sorry, but I don't understand your comment - there already is support for absl::string_view in GoogleTest.

**Comment by Jonny007-MKD** [comment]
@dermojo: In this case std::string_view is also already supported, because absl::string_view == std::string_view if available.

**Comment by dermojo** [comment]
@Jonny007-MKD: So if I understand correctly, you're requesting to merge the support code for both string_view types? That would be totally reasonable :)
I just need to make sure that all conditional compilations still work (with/without ABSL, pre/post-C++17).

**Comment by dermojo** [comment]
@Jonny007-MKD : I updated the implementatio to alias either absl::string_view or std::string_view into the ::testing namespace, so that the rest of the code doesn't need to distinguish. Hope this now looks as expected.

File: googletest/include/gtest/gtest-printers.h [hunk 1]
```diff
@@ -117,8 +117,28 @@
...
+#if GTEST_HAS_STD_STRINGVIEW
+#include <string_view>
+#endif  // GTEST_HAS_STD_STRINGVIEW
+
 namespace testing {
 
+// create an alias for string_view, wherever it comes from
+#if GTEST_HAS_ABSL
+
+using absl::string_view;
```
Comment (shahzadlone) [inline googletest/include/gtest/gtest-printers.h]: This is a neat trick! but is this a good idea to do this in a .h file ? hope this doesn't cause future contributors confusion or additional linking problems.

Comment (dermojo) [reply googletest/include/gtest/gtest-printers.h]: That saves a lot of `#if's`, but I get your point. How about assigning the typedef to a dedicated name? E.g.:
`using gtest_string_view = /*...*/`

Comment (shahzadlone) [reply googletest/include/gtest/gtest-printers.h]: perhaps that would be better

**Comment by dermojo** [comment]
@shahzadlone Updated the typedef to use `StringView` instead. This should avoid any confusion (I hope)

**Comment by kamolsrisan** [comment]
Good

**Comment by Oooocean** [comment]
#-   

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by dermojo** [comment]
Related issue: #907

**Comment by davidzchen** [comment]
I am still getting the same issue when comparing two `absl::string_view` instances. Also added a comment on #2437

**Pull request closed without merging**